// gqlx-apollo-express-server/dist/subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSubscription = void 0;
const subscriptions_transport_ws_1 = require("subscriptions-transport-ws");
const graphql_1 = require("graphql");
const context_1 = require("./context");
const constants_1 = require("./constants");
function getUpdater(subscription) {
    return (schema) => {
        subscription.schema = schema;
    };
}
function createSubscription(server, options) {
    const { paths = {}, schema, services, createApi = constants_1.defaultApiCreator } = options;
    const path = paths.subscriptions || constants_1.defaultSubscriptionsPath;
    const subscription = new subscriptions_transport_ws_1.SubscriptionServer({
        execute: graphql_1.execute,
        subscribe: graphql_1.subscribe,
        keepAlive: options.keepAlive,
        schema: schema.get(),
        async onOperation(_message, params, ws) {
            const req = ws.upgradeReq;
            const context = (0, context_1.createContext)(req, services, createApi);
            return Object.assign(Object.assign({}, params), { context });
        },
    }, {
        server,
        path,
    });
    const stopListening = schema.onUpdate(getUpdater(subscription));
    return () => {
        stopListening();
        subscription.close();
    };
}
exports.createSubscription = createSubscription;
